### PR TITLE
Sync metadata implementation with DataAPI.jl 1.12.0

### PR DESCRIPTION
**Pull request by bkamins**
This PR should be passed through CI after https://github.com/JuliaData/DataAPI.jl/pull/53 is merged and DataAPI.jl 1.12.0 is released. Other than that it should be complete (not tested, but the changes are minor - mostly simplification of the code because of dropped default implementations + adding support for `default`).

### CODE REVIEW

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:116]:
```suggestion
metadata(x::Union{DataFrameRows, DataFrameColumns}, key::AbstractString,
         default=MetadataMissingDefault(); style::Bool=false) =
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:350]:
```suggestion
If `default` is passed then return it if `key` does not exist for column `col`;
if `style=true` return `(default, :default)`.
If `col` does not exist in `df` always throw an error.
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:346]:
```suggestion
`SubDataFrame` and `DataFrameRow` expose only `:note`-style metadata of their parent.
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:89]: I don't think this precision is needed, as non-`:note`-style metadata is not exposed when `default` is provided either.
```suggestion
parent.
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:92]:
```suggestion
If `default` is passed then return it if `key` does not exist;
if `style=true` return `(default, :default)`.
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:387]:
```suggestion
    colmetadata(parent(x), col, key, default; style=style)
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:382]:
```suggestion
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:428]:
```suggestion
    idx = index(df)[col] # check if column exists and get its integer index
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:490]:
```suggestion
    idx = index(df)[col] # check if column exists and get its integer index
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:513]:
```suggestion
function colmetadata!(x::Union{DataFrameRow, SubDataFrame},
                      col::ColumnIndex, key::AbstractString, value::Any; style)
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:523]:
```suggestion
        haskey(cols_meta[idx], key) && cols_meta[idx][key][2] !== :note
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:551]:
```suggestion
    idx = index(df)[col] # check if column exists and get its integer index
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:628]:
```suggestion
    idx = index(df)[col] # check if column exists and get its integer index
```

File: src/other/metadata.jl
Comment (nalimilan) [inline src/other/metadata.jl:732]: Use an `if` for consistency with above?

File: test/metadata.jl
Comment (nalimilan) [inline test/metadata.jl:27]:
```suggestion
@testset "metadatasupport" begin
```

**Review by nalimilan — APPROVED** [review]

**Comment by nalimilan** [comment]
Nice cleanup!

File: src/other/metadata.jl
Comment (bkamins) [reply src/other/metadata.jl]: OK. I have changed short-circut to `if`s everywhere then. After CI passes I will merge this PR.
Then I will wait one day to wait for final responses, and will make a release of DataFrames.jl 1.4 tomorrow if there are none.

File: src/other/metadata.jl
Comment (bkamins) [inline src/other/metadata.jl:4]:
```suggestion
# to detect the fact that no default was passed
```

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**
